### PR TITLE
Don't create nested planner nodes when they're statically unreachable

### PR DESCRIPTION
**Pull request by dcostaras**

### CODE REVIEW

File: src/main/com/wsscode/pathom3/connect/planner.cljc [hunk 1]
```diff
@@ -910,7 +914,8 @@
                         (assoc
                           ::attr-resolvers-trail (into (or attr-resolvers-trail #{}) resolvers)
                           ::available-data available
-                          :edn-query-language.ast/node (pfsd/shape-descriptor->ast missing))))]
+                          :edn-query-language.ast/node (pfsd/shape-descriptor->ast missing)
+                          :com.wsscode.pathom3.error/lenient-mode? lenient-mode?)))]
```
Comment (wilkerlucio) [inline src/main/com/wsscode/pathom3/connect/planner.cljc:918]: For `shape-reachable?` we can always use lenient mode `true`

File: src/main/com/wsscode/pathom3/connect/planner.cljc [hunk 2]
```diff
@@ -920,6 +925,16 @@
           shape))
       true)))
 
+(>defn shape-statically-unreachable?
```
Comment (wilkerlucio) [inline src/main/com/wsscode/pathom3/connect/planner.cljc:928]: not sure if we need this function, seems a bit just a reverse of `shape-reachable?`

File: src/main/com/wsscode/pathom3/connect/planner.cljc [hunk 3]
```diff
@@ -1091,42 +1106,44 @@
...
-         :children (if sub
-                     (let [ast' (pfsd/shape-descriptor->ast sub)]
-                       [(assoc ast
-                          :children (:children ast')
-                          :query (pfsd/shape-descriptor->query sub))])
-                     [ast])}))))
+  (let [{:keys [::pco/provides] :as config} (pci/resolver-config env op-name)
+        resolver-provided-shape             (get provides (:key ast))
+        ast-required-shape                  (pfsd/ast->shape-descriptor ast)]
+    (when-not (shape-statically-unreachable? env resolver-provided-shape ast-required-shape)
```
Comment (wilkerlucio) [inline src/main/com/wsscode/pathom3/connect/planner.cljc:1112]: here is important to check if both the nested query AND the resolver sub-query have details (something not empty inside). If any is blank (the sub-query or the resolver attribute output subquery) we should short-circuit and accept the node.

File: test/com/wsscode/pathom3/connect/planner_test.cljc [hunk 4]
```diff
@@ -336,6 +336,124 @@
                                           :key          :b,
                                           :query        '...}}}))))))
 
+
+(deftest compute-run-graph--statically-unreachable-nodes-not-created
+  (testing "Nodes with nested provides that are known to be statically unreachable given nested query requires are not created"
+    (is (= #:com.wsscode.pathom3.connect.planner
```
Comment (wilkerlucio) [inline test/com/wsscode/pathom3/connect/planner_test.cljc:342]: to keep things standard, can you please swipe the position and put `(= GIVEN EXPECTED)` please?

File: test/com/wsscode/pathom3/connect/planner_test.cljc [hunk 5]
```diff
@@ -336,6 +336,124 @@
...
+               ::pcp/root 2}
+              (compute-run-graph
+               {::eql/query [{:container [:derived]}]
+                ::resolvers '[{::pco/op-name users
+                               ::pco/output  [{:container {:users [:id]}}]}
+                              {::pco/op-name derived
+                               ::pco/input   [{:users [:id]}]
+                               ::pco/output  [:derived]}]})))))
+
+(comment
```
Comment (wilkerlucio) [inline test/com/wsscode/pathom3/connect/planner_test.cljc:419]: ✂️

File: test/com/wsscode/pathom3/connect/planner_test.cljc [hunk 6]
```diff
@@ -336,6 +336,124 @@
...
+                  (pco/resolver
+                   'derived
+                   {::pco/input  [{:users [:id]}]
+                    ::pco/output [:derived]}
+                   (fn [_ _] {:derived :derived-value}))])
+                ((requiring-resolve 'com.wsscode.pathom.viz.ws-connector.pathom3/connect-env)
+                 {:com.wsscode.pathom.viz.ws-connector.core/parser-id "Root derived"}))]
+    (pcp/compute-run-graph env)))
+
+(comment
```
Comment (wilkerlucio) [inline test/com/wsscode/pathom3/connect/planner_test.cljc:436]: ✂️

File: test/com/wsscode/pathom3/connect/planner_test.cljc [hunk 7]
```diff
@@ -336,6 +336,124 @@
...
+                                                :dispatch-key :container
+                                                :key          :container
+                                                :query        [:item-1]
+                                                :children     [{:type         :prop
+                                                                :dispatch-key :item-1
+                                                                :key          :item-1}]}}
+            :index-resolver->nodes {'nested-1-resolver #{1}}
+            :index-attrs           {:container #{1}}
+            :root                  1}
+           (compute-run-graph
```
Comment (wilkerlucio) [inline test/com/wsscode/pathom3/connect/planner_test.cljc:357]: would be nice to have another case here where there are multiple options, but user doesn't send the sub-query (so keep both).

Also, the case where the user send a sub-query, but the output doesn't have any nested detail (should also keep all options in this case).

**Review by wilkerlucio — CHANGES_REQUESTED** [review]
Hello, I add some changes requests, but overall looking good 👍

File: src/main/com/wsscode/pathom3/connect/planner.cljc (hunk 1 above)
Comment (dcostaras) [reply src/main/com/wsscode/pathom3/connect/planner.cljc:918]: Cool, that makes sense though I think that changes current behaviour if I remember correctly. I'll double check and get back to you

File: src/main/com/wsscode/pathom3/connect/planner.cljc (hunk 3 above)
Comment (dcostaras) [reply src/main/com/wsscode/pathom3/connect/planner.cljc:1112]: This check is done in `shape-statically-unreachable?`. I'll move it here and get rid of the extra function.

File: test/com/wsscode/pathom3/connect/planner_test.cljc (hunk 4 above)
Comment (dcostaras) [reply test/com/wsscode/pathom3/connect/planner_test.cljc:342]: Will do

File: test/com/wsscode/pathom3/connect/planner_test.cljc (hunk 6 above)
Comment (dcostaras) [reply test/com/wsscode/pathom3/connect/planner_test.cljc:436]: Will do, just me debugging still

File: test/com/wsscode/pathom3/connect/planner_test.cljc (hunk 7 above)
Comment (dcostaras) [reply test/com/wsscode/pathom3/connect/planner_test.cljc:357]: Thanks, those are both very useful cases

File: src/main/com/wsscode/pathom3/connect/planner.cljc (hunk 3 above)
Comment (dcostaras) [reply src/main/com/wsscode/pathom3/connect/planner.cljc:1112]: I may well have misunderstood things but it seemed like there is behaviour that relies on `shape-unreachable?` throwing in order to short circuit processing. Again, let me investigate a bit more and I'll get back to you.

File: test/com/wsscode/pathom3/connect/planner_test.cljc [hunk 8]
```diff
@@ -2205,15 +2338,15 @@
                                       ::pco/input   [:b]
                                       ::pco/output  [:c]}]
               ::eql/query           [{:a [:c]}]})
-           {::pcp/nodes                 {2 {::pco/op-name        'dyn
```
Comment (dcostaras) [inline test/com/wsscode/pathom3/connect/planner_test.cljc:2208]: Node number changes I would personally consider implementation details?

File: test/com/wsscode/pathom3/connect/runner_test.cljc [hunk 9]
```diff
@@ -1021,7 +1021,7 @@
           (graph-response?
             (pci/register
               [(pco/resolver 'users
-                 {::pco/output [{:main-db {:users [:user/id]}}]}
+                 {::pco/output [{:main-db [{:users [:user/id]}]}]}
```
Comment (dcostaras) [inline test/com/wsscode/pathom3/connect/runner_test.cljc:1024]: This took me a long time to figure out as once I had introduced the change this test started failing and I didn't consider the test to be broken. Once I had a good grasp of the planner code I thought there was a bug in the EQL AST! Eventually I realised the test was wrong... shows I don't have a lot of experience looking at EQL queries else I would have probably picked it up sooner!

Comment (wilkerlucio) [reply test/com/wsscode/pathom3/connect/runner_test.cljc:1024]: Good catch, the EQL was valid if the test was about Unions, but seems like not the case here 👍

File: test/com/wsscode/pathom3/connect/planner_test.cljc (hunk 8 above)
Comment (wilkerlucio) [reply test/com/wsscode/pathom3/connect/planner_test.cljc:2208]: Yes, no worries about that

File: test/com/wsscode/pathom3/connect/planner_test.cljc [hunk 10]
```diff
@@ -621,7 +754,7 @@
 
       (is (thrown-with-msg?
             #?(:clj Throwable :cljs js/Error)
-            #"Pathom can't find a path for the following elements in the query: \[:c]"
```
[the remark this answers is not in the document]
Comment (wilkerlucio) [reply test/com/wsscode/pathom3/connect/planner_test.cljc:624]: I think that may be better, lets give time to it and see how debuggable it gets.

**Comment by wilkerlucio** [comment]
Looking good, you can run `bb format-fix` to fix the code style.

Also some tests are breaking on CLJS, this is usually due to small differences in map orders between CLJ/CLJS, which end up making different node orders. In those case we need to use cljc reader to make different numbers for CLJ/CLJS.

Can you fix those? Other than that looks good to merge 👍

**Comment by dcostaras** [comment]
Sure thing, sorry I didn't even notice the failing CI checks!

**Comment by dcostaras** [comment]
@wilkerlucio apologies for not getting to this, it's been a mad few weeks. I will address as soon as I can 👍

**Comment by wilkerlucio** [comment]
Closing this, the latest commit implements this feature: a5424450fafee1c85f6e414cfd8b50ec09eb2a5b

**Pull request closed without merging**